hydra_pywr/exporter.py

```diff
@@ -218,11 +237,90 @@ def handler(event, args):
                 fp.write(rule["value"])
                 fp.write("\n\n")
 
+    def get_external_files(self):
+        """
+            Request the locations of any external files from hydra, using the project appdata column, 
+            then replace the relevant file names in the 'url' section of parameters and tables with the full path to the file.
+            for example replace:
+            {
+            ...
+            "url" : "demand.csv"
+            ...
+            }
+            with
+            {
+            ...
+            "url": "/path/to/demand.csv"
+            ...
+            }
+            where '/path/to' is defined in the project's metadata (the appdata column)
+        """
+        log.info("Retrieving external files")
+        #First get the project hierarchy
+        project_hierarchy = self.hydra.get_project_hierarchy(project_id=self.data['project_id'])
+
+        #start from the top down. Files with the same neame, at a lower level
+        #take precedence.
+        project_hierarchy.reverse()
+
+        self.filedict = {}
+        self.fileprojects  = {}
+        for proj_in_hierarchy in project_hierarchy:
+            #Files uploaded to the USER_FILE_UPLOAD_DIR are synchronized with the USER_FILE_ROOT_DIR
+            #So they are then downloaded from the USER_FILE_ROOT_DIR
+            if proj_in_hierarchy.appdata is None:
+                continue
+
+            data_s3_bucket = proj_in_hierarchy.appdata.get('data_s3_bucket')
+
+            if data_s3_bucket is None:
+                return
+
+            project_data_path = proj_in_hierarchy.appdata.get('data_uuid')
+
+            if project_data_path is None:
+                continue
+
+            local_project_data_dir = self.sync_with_s3(data_s3_bucket, project_data_path) #TODO: How do we do this?
+
+            try:
+                projectfiles = os.listdir(local_project_data_dir)
+                for f in projectfiles:
+                    name = f.split(os.sep)[-1]
+                    self.filedict[name] = os.path.join(local_project_data_dir, f)
+                    self.fileprojects[name] = proj_in_hierarchy
+            except FileNotFoundError:
+                pass
+        log.info("External files synced")
+        return self.filedict
+
+
+    def sync_with_s3(self, s3_bucket_name, project_data_path):
+        """
+            Sync the data folder with the specified s3 bucket
+        """
+        log.info(f"Syncing with s3 bucket {s3_bucket_name}")
+
+        data_dir = os.path.join(self.data_dir, 'projectdata', project_data_path)
+
+        sync_command = f"aws s3 sync s3://{s3_bucket_name}/data/projectdata/{project_data_path} {data_dir}"
```

Comment:
Is `aws` always guaranteed to be in the shell path?

There is also the per-file `s3fs.S3FileSystem` approach which supports authentication, see `utils.retrieve_s3`.